File: java/20.java
class Solution {
    public boolean isValid(String s) {
        char stack[] = new char[s.length()];
        int i = -1;
        for (char c : s.toCharArray())
            if (c == '(' || c == '{' || c == '[')
                stack[++i] = c;
            else {
                if (i == -1 || (c == ')' && stack[i] != '(') || (c == '}' && stack[i] != '{')
                        || (c == ']' && stack[i] != '['))
                    return false;
                --i;
            }
        return i == -1;
    }
}

File: cpp/20.cpp
class Solution {
public:
    bool isValid(string s) {
        int n = s.length(), i = -1;
        char* stack = (char*)malloc(sizeof(char) * n);
        for (int j = 0; j < n; j++)
            if (s[j] == '(' || s[j] == '{' || s[j] == '[')
                stack[++i] = s[j];
            else {
                if (i == -1 || (s[j] == ')' && stack[i] != '(') ||
                    (s[j] == '}' && stack[i] != '{') ||
                    (s[j] == ']' && stack[i] != '['))
                    return false;
                --i;
            }
        return i == -1;
    }
};
